text/0000-lifecycle-parallel-export.md

```diff
@@ -0,0 +1,182 @@
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is to add a new capability to the lifecycle (enabled by configuration) to export app image and cache image to registry in parallel.
+
+The target personas affected by this change are:
+
+ - **buildpack user**: they will experience higher disk I/O or network pressure if this feature is enabled by default.
+ - **Platform implementors**: they will choose parallel or serial export, to suit how the platform works. Serial export helps to get app image faster, while parallel export can complete the build process faster.
+
+This proposal, in addition to acceleration, has the greatest impact on the export time of app image. This will lead to resource competition, that is, when the app image and cache image are exported at the same time, the app image export time will become longer. For some users, they only care about the export time of the app image, but not the overall optimization effect, and enabling this capability will affect their performance. Therefore, this ability is optional.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+It will be done using goroutines. Goroutine is a lightweight multi-threading mechanism, which can avoid a lot of extra overhead caused by the multi-threaded parallel operation. This function encapsulates the export process of the app image and the cache image into two goroutines to execute in parallel.
+
+The working principle is shown in the following code (go-like pseudocode). Execute in parallel through two goroutines and wait for all export processes to finish. If parallel export is not enabled, the process is exactly the same as the original serial export.
+
+```go
+func export() {
+    exporter := &lifecycle.Exporter{}
+    // ...
+    
+    var wg sync.WaitGroup
+    
+    wg.Add(1)
+    go func() {
+        defer wg.Done()
+        exporter.Export()
+    }()
+	
+    if !enableParallelExport {
+        wg.Wait()
+    }
+    
+    wg.Add(1)
+    go func() {
+        defer wg.Done()
+        exporter.Cache()
+    }()
+    
+    wg.Wait()
+    
+    // ...
+}
+```
+
+## Examples
+
+For command line use, control this process through the `parallel` flag.
+
+### Export both app image and cache image
+
+By specifying environment variable `CNB_PARALLEL_EXPORT`, or pass a `-parallel` flag, images will be pushed to `cr1.example.com` and `cr2.example.com` simultaneously.
+
+```shell
+> export CNB_PARALLEL_EXPORT=true
+> /cnb/lifecycle/exporter -app cr1.example.com/foo:app -cache-image cr2.example.com/foo:cache
+
+# OR
+
+> /cnb/lifecycle/exporter -app cr1.example.com/foo:app -cache-image cr2.example.com/foo:cache -parallel
+```
+
+### Export app image only or export cache image only
+
+If export one image, the effect of this function is not very obvious.
+
+```shell
+> export CNB_PARALLEL_EXPORT=true
+> /cnb/lifecycle/exporter -app cr1.example.com/foo:app
+
+# OR
+
+> /cnb/lifecycle/exporter -app cr1.example.com/foo:app -parallel
+
```

Comment:
I would suggest adding a `WARN` message to the end-user to indicate that using the `parallel` flag without exporting a cache image has no effect.

Reply:
This makes sense, I agree with you.